[SPARK-39883][SQL][TESTS] Add DataFrame function parity check

aray opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR adds a test that compares the available list of DataFrame functions in org.apache.spark.sql.functions with the SQL function registry. This attempts to verify that the DataFrame functions are a subset of the functions in the SQL function registry (subject to exclusions and expectations). ~It also produces a list of the differences between the two.~(removed per feedback)

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Currently there is no easy way to verify what the difference is between the two API's 

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
This PR is testing only

aray committed:
port from other branch

aray committed:
revisions
* add dev notes
* expectations assert
* report with println all at end

aray commented:
@cloud-fan or @MaxGekk can you review this or suggest who else might? For comparison see also https://github.com/apache/spark/pull/37144 for PySpark function parity check that is already merged.

cloud-fan commented:
looks fine to me, cc @HyukjinKwon @viirya

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -41,6 +42,100 @@ import org.apache.spark.sql.types._
+    println(s"  There are ${dataFrameFunctions.size} relevant functions in the DataFrame API")
+    println(s"  There are ${sqlFunctions.size} relevant functions in the SQL function registry")
+    println(s"  Number of functions in both sets: $commonCount")
+    if(onlyDataFrameFunctions.nonEmpty) {
+      val number = onlyDataFrameFunctions.size
+      val sortedList = onlyDataFrameFunctions.toList.sorted.mkString(", ")
+      println(s"  There are $number DataFrame functions that are not in SQL: $sortedList")
+    }
```
As this is matched expectation, do we need to print it out?

aray replied:
Sure, I'll remove this section

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -41,6 +42,100 @@ import org.apache.spark.sql.types._
+    assert(onlySqlFunctions.intersect(excludedDataFrameFunctions) === Set.empty)
+
+    // Check that only expected functions are left
+    assert(onlyDataFrameFunctions === expectedOnlyDataFrameFunctions)
```
In error message, maybe showing what are not expected to be only in DataFrame functions?

aray replied:
I'm updating this to provide the symmetric difference as a clue. For example a failure would then look like: `Set("years", "hours", "unwrap_udt", "bucket", "months", "days", "product") did not equal Set("years", "demo_missing", "hours", "unwrap_udt", "bucket", "months", "days", "product") symmetric difference is: Set(demo_missing)`

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -41,6 +42,100 @@ import org.apache.spark.sql.types._
+      val sortedList = onlyDataFrameFunctions.toList.sorted.mkString(", ")
+      println(s"  There are $number DataFrame functions that are not in SQL: $sortedList")
+    }
+    if(onlySqlFunctions.nonEmpty) {
```
```suggestion
    if (onlySqlFunctions.nonEmpty) {
```

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -41,6 +42,100 @@ import org.apache.spark.sql.types._
+    println(s"  There are ${dataFrameFunctions.size} relevant functions in the DataFrame API")
+    println(s"  There are ${sqlFunctions.size} relevant functions in the SQL function registry")
+    println(s"  Number of functions in both sets: $commonCount")
+    if(onlyDataFrameFunctions.nonEmpty) {
```
```suggestion
    if (onlyDataFrameFunctions.nonEmpty) {
```

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -41,6 +42,100 @@ import org.apache.spark.sql.types._
+    assert(onlyDataFrameFunctions === expectedOnlyDataFrameFunctions)
+
+    // scalastyle:off println
+    println("Report: DataFrame function and SQL functon parity")
```
I'm not sure if it is necessary to print out these info. Normally we only notice test output when it is failed.

aray replied:
Since we don't have any explicit check on the large list of functions that are only available in SQL, I though it would be useful for developers to produce a report of the functions that are only available there. Granted they will need to explicitly run this single test to get that report in a usable fashion. For example this report (after pr feedback) will look like: 

Report: DataFrame function and SQL functon parity
  There are 228 relevant functions in the DataFrame API
  There are 366 relevant functions in the SQL function registry
  Number of functions in both sets: 221
  There are 145 SQL functions that are not in the DataFrame API: aes_decrypt, aes_encrypt, and, any, any_value, array_size, bigint, binary, bit_and, bit_count, bit_get, bit_or, bit_xor, bool_and, bool_or, boolean, btrim, cast, char, chr, contains, convert_timezone, count_if, count_min_sketch, current_catalog, current_database, current_timezone, current_user, date, date_from_unix_date, date_part, decimal, div, double, e, elt, endswith, equal_null, every, extract, find_in_set, float, histogram_numeric, if, ifnull, ilike, in, inline, inline_outer, input_file_block_length, input_file_block_start, int, isnotnull, json_array_length, json_object_keys, left, like, ln, make_dt_interval, make_interval, make_timestamp, make_timestamp_ltz, make_timestamp_ntz, make_ym_interval, median, mod, mode, named_struct, now, nullif, nvl, nvl2, or, parse_url, percentile, pi, positive, power, regexp, regexp_count, regexp_extract_all, regexp_instr, regexp_like, regexp_substr, regr_avgx, regr_avgy, regr_count, regr_intercept, regr_r2, regr_slope, regr_sxx, regr_sxy, regr_syy, replace, right, rlike, smallint, some, space, split_part, stack, startswith, str_to_map, string, timestamp, timestamp_micros, timestamp_millis, tinyint, to_binary, to_char, to_number, to_timestamp_ltz, to_timestamp_ntz, to_unix_timestamp, try_add, try_avg, try_divide, try_element_at, try_multiply, try_subtract, try_sum, try_to_binary, try_to_number, try_to_timestamp, typeof, unix_date, unix_micros, unix_millis, unix_seconds, url_decode, url_encode, user, uuid, version, weekday, width_bucket, xpath, xpath_boolean, xpath_double, xpath_float, xpath_int, xpath_long, xpath_number, xpath_short, xpath_string

HyukjinKwon replied:
Hm, I think we should better avoid printing out too actually. It is useful but I think the purpose of this is to test. Can we report these info when there's a test failure?

aray committed:
pr feedback

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -41,6 +42,98 @@ import org.apache.spark.sql.types._
+    // NOTE FOR DEVELOPERS:
+    // If this test fails one of the following needs to happen
+    // * If a function was added to org.apache.spark.sql.functions but not the function registry
+    //     add it to the below expectedOnlyDataFrameFunctions set.
```
It is nice to know the feature gap but this case actually happens pretty often. `org.apachep.spark.sql.functions._` will either have the commonly used functions or some feature parity w/ other libraries (e.g., pandas) as documented in that file.
cc @rxin

aray replied:
To clarify this test only checks that `org.apachep.spark.sql.functions._` is a subset of functions defined in the SQL function registry (subject to exclusions and expectations), not the other way around. In the other direction we simply provide a report of functions that are only available in the SQL function registry.

aray committed:
Merge remote-tracking branch 'origin/master' into scala-function-parity-test

HyukjinKwon reviewed (approved):
LGTM otherwise

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -41,6 +42,98 @@ import org.apache.spark.sql.types._
+    val onlySqlFunctions = sqlFunctions.diff(dataFrameFunctions)
+
+    // Check that we did not incorrectly exclude any functions leading to false positives
+    assert(onlyDataFrameFunctions.intersect(excludedSqlFunctions) === Set.empty)
```
`onlyDataFrameFunctions.intersect(excludedSqlFunctions).isEmpty`?

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -41,6 +42,98 @@ import org.apache.spark.sql.types._
+    assert(onlyDataFrameFunctions.intersect(excludedSqlFunctions) === Set.empty)
+    assert(onlySqlFunctions.intersect(excludedDataFrameFunctions) === Set.empty)
+
+    def symmDiff[T](a: Set[T], b: Set[T]): Set[T] = a.union(b).diff(a.intersect(b))
```
I personally just inline this function since this is only used once. but don't feel strongly.

aray committed:
isEmpty and inline symmDiff

aray committed:
Merge master

aray committed:
dont print report and update expectation

aray commented:
@viirya @HyukjinKwon I've addressed both of your comments including removing the printed report. I've also merged master and updated the expectations for one added function.

viirya reviewed (commented):
Looks okay. I'd leave this to @HyukjinKwon or @cloud-fan.

HyukjinKwon commented:
Merged to master.
